### PR TITLE
Add task_wrapper for for_each, for_each_index, transform and find_if …

### LINKED ISSUE #504: Per chunk RAII support
**Issue Summary:**

We are encountering a problem related to managing floating point exceptions (FPU) on the CPU. Currently, we are using a workaround involving a lambda function to maintain FPU settings across subtasks, but this approach is both costly and inefficient. Instead, we need to implement a default chunk wrapper that applies FPU settings at the start of a task and restores the previous FPU state at the task's end. This change will optimize performance and prevent corruption of tasks that yield to the executor.

**Problem Description:**

We need to manage the FPU settings for CPU operations in our code. The challenge is ensuring that all subtasks created after setting the FPU context inherit the parent task's FPU settings. Currently, we achieve this by encapsulating the loop body within a lambda function. This lambda function captures the FPU state before calling `executor.corun` and then applies this FPU state on every iteration of the loop. However, this approach has proven to be inefficient and resource-intensive.

The inefficiency arises from the overhead associated with making these FPU state calls on every iteration of the loop. Moreover, this approach is unnecessary as FPU settings only need to be applied at the beginning of each chunk or task. At the end of each task, we must also restore the previous FPU state to avoid corrupting any tasks that have yielded to the executor at this point.

**Proposed Solution:**

To address this problem efficiently, we propose implementing a default chunk wrapper for all algorithms. This wrapper would take care of applying the necessary FPU settings at the beginning of each chunk or task, and it would ensure the restoration of the previous FPU state at the task's end. By doing so, we can eliminate the overhead and inefficiency associated with applying FPU settings on every loop iteration, resulting in improved performance and avoiding task corruption issues.

**Desired Outcome:**

Implement a default chunk wrapper that handles FPU settings at the start and end of tasks, optimizing performance and preventing task corruption issues. This change will simplify code and improve overall efficiency in managing floating point exceptions on the CPU.

**Psuedo Code**

```
auto defaultChunkWrapper = [](auto && chunkRunner){
    std::forward<decltype(chunkRunner)>(chunkRunner());
};
```

however in our case we want something like

```
FPUSaver saver;
auto fpuChunkWrapper = [&saver](auto && chunkRunner){
    // Appy stored FPU settings to the current tasks and
    // restore the FPU setting as the end of the task
    FPUApplier applier(saver);
    std::forward<decltype(chunkRunner)>(chunkRunner());
};

Taskflow taskflow;
taskflow.for_each_index(0,10000,1,[](double i){  
     // trigger FP exception
     i / 0.0;
}, tf::DynamicPartioner(), chunkRunner);

executor.corun(taskflow);
```

The global https://taskflow.github.io/taskflow/classtf_1_1ObserverInterface.html interface doesn't really solve the problem in an elegant way.

### PR DESCRIPTION
**Pull request by bradphelan**
fixes #504 

This can be used for multiple purposes. For handling exceptions we could do

```cpp
tf::Executor executor(tc);
std::exception_ptr x;
auto task_wrapper = [&](auto &&task)
{
  try {
  task();
  }catch(...){
    x = std::current_exception();
  }
};
std::atomic<int> count = 0;
tf::Taskflow taskflow;
taskflow.for_each_index(
    0, 100, 1, [&](int i)
    { count++; },
    tf::StaticPartitioner(1), task_wrapper);
executor.run(taskflow).wait();
if(x)
   std::rethrow_exception(exceptionPtr);
```

Our use case is to apply captured floating point settings from parent tasks to sub tasks 

```
class FPUSaver;
class FPUApplier;
```

```cpp
tf::Executor executor(tc);
std::exception_ptr x;
/// Capture the floating point unit configuration
FPUSaver saver;
auto task_wrapper = [&](auto &&task)
{
  try {
    /// Apply the floating point unit configuration ( and undo on destruction
    FPUApplier applier(saver);
    task();
  }catch(...){
    x = std::current_exception();
  }
};
std::atomic<int> count = 0;
tf::Taskflow taskflow;
taskflow.for_each_index(
    0, 100, 1, [&](int i)
    { count++; },
    tf::StaticPartitioner(1), task_wrapper);
executor.run(taskflow).wait();
if(x)
   std::rethrow_exception(exceptionPtr);

```

### CODE REVIEW

File: taskflow/algorithm/for_each.hpp [hunk 1]
```diff
@@ -36,16 +38,18 @@ TF_FORCE_INLINE auto make_for_each_task(B beg, E end, C c, P&& part = P()) {
...
-              std::advance(beg, curr_b - prev_e);
-              for(size_t x = curr_b; x<curr_e; x++) {
-                c(*beg++);
+        launch_loop(W, w, rt, [=, &c, &part,&task_wrapper] () mutable {
+          task_wrapper([&]{
+            part.loop(N, W, curr_b, chunk_size,
+              [&, prev_e=size_t{0}](size_t curr_b, size_t curr_e) mutable {
+                std::advance(beg, curr_b - prev_e);
+                for(size_t x = curr_b; x<curr_e; x++) {
+                  c(*beg++);
```
Comment (olologin) [inline taskflow/algorithm/for_each.hpp:47]: Shouldn't we execute wrapper for this thread?

File: taskflow/algorithm/for_each.hpp [hunk 2]
```diff
@@ -54,32 +58,34 @@ TF_FORCE_INLINE auto make_for_each_task(B beg, E end, C c, P&& part = P()) {
...
-            std::advance(beg, curr_b - prev_e);
-            for(size_t x = curr_b; x<curr_e; x++) {
-              c(*beg++);
+      launch_loop(N, W, rt, next, part, [=, &c, &next, &part,&task_wrapper] () mutable {
+        task_wrapper([&]() {
+          part.loop(N, W, next, 
+            [&, prev_e=size_t{0}](size_t curr_b, size_t curr_e) mutable {
+              std::advance(beg, curr_b - prev_e);
+              for(size_t x = curr_b; x<curr_e; x++) {
+                c(*beg++);
```
Comment (olologin) [inline taskflow/algorithm/for_each.hpp:67]: Shouldn't we execute wrapper here?

File: taskflow/algorithm/for_each.hpp (hunk 1 above)
Comment (olologin) [reply taskflow/algorithm/for_each.hpp:47]: I mean for general case, nobody knows if user-supplied wrapper can be skipped for parent thread.

**Comment by olologin** [comment]
I resolved merge conflicts and rebased it on the current master. Also fixed problems that I highlighted before.

@tsung-wei-huang could this be reviewed and merged? We are using this patch internally and it helps us with performance for situations when tasks have to be executed in some environment with specific FPU settings, etc.

**Comment by tsung-wei-huang** [comment]
@olologin I am reviewing it today and will get it merged if no problem :)

**Comment by olologin** [comment]
Hmm, I will look into these two failing tests.

**Comment by olologin** [comment]
Tests should be fixed now. But TSAN job failure seems like something that is not caused by our change. Correct me if I am wrong.

**Comment by tsung-wei-huang** [comment]
I am thinking to combine TaskWrapper with Partitioner, which allows more flexible extension in the future using Partitioner. For example:

```
tf::Executor executor(tc);
std::exception_ptr x;
tf::StaticPartitioner part(1, [&](auto &&task){
  try {
  task();
  }catch(...){
    x = std::current_exception();
  }
});

std::atomic<int> count = 0;
tf::Taskflow taskflow;
taskflow.for_each_index(
    0, 100, 1, [&](int i)
    { count++; },
    tf::StaticPartitioner(1)
);
executor.run(taskflow).wait();
if(x)
   std::rethrow_exception(exceptionPtr);
```

What do you think?

**Comment by bradphelan** [comment]
I'm not sure that the coupling is correct. For example there are algorithms that you provide that do not take a partitioner such a parallel scan but still require a task wrapper. 

image

**Comment by tsung-wei-huang** [comment]
It depends on the algorithm itself. parallel-scan does not support partitioner due to its algorithm nature.

**Comment by bradphelan** [comment]
> It depends on the algorithm itself. parallel-scan does not support partitioner due to its algorithm nature.

That is my point. Not all algorithms use partitioners but you are proposing integrating a feature with partitioners which would make the feature unavailable to algorithms that do not use partitioners.

**Comment by tsung-wei-huang** [comment]
@bradphelan @olologin thank you for contributing to this great feature. I have merged it - as I said, don't worry about the c++20 version of the failed test. I will fix it later.

**Pull request merged**